metrics/network/cpu_statistics_iperf.sh

```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+#
+# Copyright (c) 2018 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# Description of the test:
+# This test runs the 'iperf' network benchmark and
+# we retrieve cpu information while running the networking
+# test with 'perf'. CPU utilization provides more stable
+# results that help to detect performace regressions compared
+# with the actual throughput.
+
+set -e
+
+# General env
+SCRIPT_PATH=$(dirname "$(readlink -f "$0")")
+source "${SCRIPT_PATH}/../lib/common.bash"
+
+# Test name
+TEST_NAME="${TEST_NAME:-cpu information}"
+# Image name
+IMAGE="${IMAGE:-local-iperf}"
+# Dockerfile
+DOCKERFILE="${SCRIPT_PATH}/iperf_dockerfile/Dockerfile"
+# Client name
+CLIENT_NAME="${CLIENT_NAME:-client_iperf}"
+# Server name
+SERVER_NAME="${SERVER_NAME:-server_iperf}"
+# Target bandwidth
+BANDWIDTH="${BANDWIDTH:-1G}"
+# Time in seconds to transmit
+TIMEOUT="${TIMEOUT:-160}"
+# Sample period in seconds to measure
+SAMPLE_PERIOD="${SAMPLE_PERIOD:-120}"
+# File to save perf results
+TMP_FILE=$(mktemp cpuinfo.XXXXXXXXXX)
+# Time in seconds is required to let the benchmark settle down
+SETTLE_TIME="${SETTLE_TIME:-20}"
+# Set the runtime if not set already
+RUNTIME="${RUNTIME:-kata-runtime}"
+
+function main() {
+	# Check tools/commands dependencies
+	cmds=("awk" "docker" "perf")
+
+	init_env
+	check_cmds "${cmds[@]}"
+	check_dockerfiles_images "$IMAGE" "$DOCKERFILE"
+
+	# Start iperf3 server configuration
+	# Set the TMPDIR to an existing tmpfs mount to avoid a 9p unlink error
+	local init_cmds="export TMPDIR=/dev/shm"
+	local server_cmd="$init_cmds && iperf -s"
+	docker run --runtime $RUNTIME --name $SERVER_NAME -d $IMAGE bash -c "$server_cmd"
+	local server_address=$(docker inspect --format "{{.NetworkSettings.IPAddress}}" $SERVER_NAME)
+
+	# Start iperf3 client
+	local client_cmd="$init_cmds && iperf -c $server_address -b $BANDWIDTH -t $TIMEOUT"
+	docker run --runtime $RUNTIME --name $CLIENT_NAME -d $IMAGE bash -c "$client_cmd"
+
+	# This time in seconds is required to let the benchmark settle down
+	sleep "$SETTLE_TIME"
+
+	# Retrieve qemu information
+	PIDS="$(pgrep -d ',' qemu)"
+
+	metrics_json_init
+
+	# Start collecting cpu information
+	sudo perf stat -o $TMP_FILE -e cycles -e instructions -p ${PIDS} sleep ${SAMPLE_PERIOD}
+
+	# Save configuration
+	metrics_json_start_array
+
+	local result=$(cat $TMP_FILE)
+	local instructions_per_cycle="$(echo "$result" | grep -w "insns per cycle" | cut -d '#' -f2 | awk '{print $1}')"
+	local cycles="$(echo "$result" | grep -w "cycles" | awk '{print $1}' | tr ',' ' ' | \
```

Comment:
I've a feeling this 'find the max' calculation is all wrong. Here is the example `pef` output I get from a local run:
```
   216,511,365,940      cycles
   373,206,617,528      instructions              #    1.72  insns per cycle
```

It looks like you are parsing off the largest 3 digit number from that comma separated sequence - **but** - that is not a sequence of numbers, that is **one number** - one very large number :-). So, that is telling you we did 373 *billion* instructions in 216 *billion* cycles.
So, no max parsing necessary - we probably need to remove the `,`'s from the results (and maybe `perf` even has an option for that?).

Please fix, re-run the test and re-post your results to check reproducibility.

Reply:
So, looking at this entire thread, it seems like a lot of time is being spent in review that shouldn't have been necessary.  I think more due diligence prior to submitting code (like testing) would be good.